### PR TITLE
Exclude keys containing empty patches in the final patch

### LINKED ISSUE #58793: kubectl apply does not properly minimize patch when using JSON merge patch
**Is this a BUG REPORT or FEATURE REQUEST?**: Bug Report
/kind bug

**What happened**: Assume a CustomResource
original:
```
{
    "apiVersion": "company/v1",
    "kind": "CustomResource",
    "metadata": {
        "annotations": {},
        "name": "secrets",
        "namespace": "test"
    }
}
```
current:
```
{
    "apiVersion": "company/v1",
    "kind": "CustomResource",
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"company/v1\",\"kind\":\"VaultSecret\",\"metadata\":{\"annotations\":{},\"name\":\"secrets\",\"namespace\":\"test\"}}\n"
        },
        "clusterName": "",
        "creationTimestamp": "2018-01-24T22:49:30Z",
        "deletionGracePeriodSeconds": null,
        "deletionTimestamp": null,
        "generation": 0,
        "name": "secrets",
        "namespace": "test",
        "resourceVersion": "4800",
        "uid": "d6e8e5d8-0158-11e8-be12-080027cce81e"
    },
    "time-processed": "2018-01-24T22:49:30"
}
```
modified:
```
{
    "apiVersion": "company/v1",
    "kind": "CustomResource",
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"company/v1\",\"kind\":\"VaultSecret\",\"metadata\":{\"annotations\":{},\"name\":\"secrets\",\"namespace\":\"test\"}}\n"
        },
        "name": "secrets",
        "namespace": "test"
    }
}
```

original add-and-change patch current -> modified
```
{
    "metadata": {
        "clusterName": null,
        "creationTimestamp": null,
        "deletionGracePeriodSeconds": null,
        "deletionTimestamp": null,
        "generation": null,
        "resourceVersion": null,
        "selfLink": null,
        "uid": null
    }
}
```
filtered add-and-change patch (non-nulls only)
Note that at this point, the patch could have been optimized to `{}`
```
{"metadata":{}}
```

original deleted patch between original -> modified
```
{
    "metadata": {
        "annotations": {
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"company/v1\",\"kind\":\"VaultSecret\",\"metadata\":{\"annotations\":{},\"name\":\"secrets\",\"namespace\":\"test\"}}\n"
        }
    }
}
```
filtered deleted patch (nulls only)
Note that at this point, the patch could have been optimized to `{}`
```
{"metadata":{"annotations":{}}}
```

This results in a non-empty patch, which causes issues since we forbid patches to most users. We'd like for unchanged files to not result in unnecessary PATCH requests, as they result in forbidden requests.

**What you expected to happen**: The modified object above should resolve to an empty patch, and kubectl should not send a PATCH request to the apiserver.

**How to reproduce it (as minimally and precisely as possible)**: create a CustomResourceDefinition, create an instance of the resource using a file, forbid patching to the custom resource, re-apply the same file - you should receive a forbidden error.

**Anything else we need to know?**: This is totally kubectl-side issue - kubectl is calculating a patch, and then sends an unnecessary PATCH request to a non-optimal patch resolution.
I have a fix ready that fixes the issue, but need to wait for my company to read over the CLA.

The issue lies with the code block here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch.go#L116
The block fails to clean up nested maps that no longer contain any patches, eg `{a: { b: nil } }` filters nils to `{a: {} }` when really it can go further and resolve to `{}`

There's 2 solutions that I can see, I chose the former:
1) Modify the linked block above to remove empty filtered maps if they weren't already explicitly set in the original map
2) Preserve the existing code, and then in kubectl's apply.go go over the filtered patch and remove empty patches there

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.6 client/server, but verified `kubectl` built off current master has the same issue.
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): N/A (kubectl running on Mac)
- Kernel (e.g. `uname -a`): N/A
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by davidhao3300**
**What this PR does / why we need it**: 
This minimizes the 3-way JSON merge patch generated when calculating the patch necessary to send to the server. It does this by removing empty maps created from deleting keys in the keepOrDeleteNullInObj method.

This is not only a slight performance improvement (less PATCH requests) but also necessary when working with custom resources that have RBAC restrictions.

**Which issue(s) this PR fixes**: N/A

**Special notes for your reviewer**: N/A

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Quentin-M** [comment]
Thank you for that.

I believe we'd need the same or similar thing for `strategicpatch.CreateThreeWayMergePatch`, but it might be unrelated. The inconsistency between `cmdutil.NewFactory(cc).NewBuilder().Unstructured().Stream()` and what the server sends back on `*resourceInfo.Get()` (when getting the live version of the object, versus the manifest object that the builder just read) is disheartening.

It happens mainly on annotations, labels across the board, and on namespace for root scoped resources. The patch would contain an empty `metadata { namespace: "" }` change. Then there's also the fact that the `strategicpatch.CreateThreeWayMergePatch` function hardcodes the `setElementOrder` diff option, which creates useless or undesired diffs: sometimes changing the order of a list of .. one element, or shuffling items (e.g. env vars, containers) because something somewhere in the pipe decided so.

Not sure how to fix these above.. It makes it hard to detect whether an actual patch has been applied or should be applied.

**Comment by deads2k** [comment]
@ironcladlou you chasing patch in your path through apply?

**Comment by ironcladlou** [comment]
@deads2k 
>@ironcladlou you chasing patch in your path through apply?

The apply fun that @juanvallejo and I were chasing didn't concern the particulars of the patch mechanics/merge strategy; it was more about clarifying validation and flag semantic inconsistencies, etc.

**Comment by davidhao3300** [comment]
all tests are passing, ready for review

**Comment by deads2k** [comment]
/assign soltysh

more lines removed than added. That's a good sign :)

**Comment by davidhao3300** [comment]
tests are passing now

**Review by soltysh — APPROVED** [review]
The cli changes are fine with me, you need to get @lavalamp 's approval though on the apimachinery changes

**Comment by lavalamp** [comment]
Is this really 100% backwards compatible? @apelisse @pwittrock Would you expect existing testing to catch any backwards compatibility problems this might cause?

**Comment by apelisse** [comment]
@mengqiy has a better understanding of SMP than I do.

**Comment by davidhao3300** [comment]
Sorry for the inactivity; I was on vacation. Now looking at test failures.

Re: backwards compatibility - not sure what our policy on testing that is like. Logically, this should only reduce the number of PATCH requests sent to the server; it shouldn't change any other behavior unless people were expecting what should be no-op PATCH requests

**Comment by davidhao3300** [comment]
guess I should've retested instead of rebasing...

**Comment by mengqiy** [comment]
The verify test is failing, I have a PR to fix it.
https://github.com/kubernetes/kubernetes/pull/62148

**Comment by mengqiy** [comment]
> Would you expect existing testing to catch any backwards compatibility problems this might cause?

Presubmit test suite can't catch that.
Only skew test setup may catch backwards compatibility issue.

**Comment by mengqiy** [comment]
> there's also the fact that the strategicpatch.CreateThreeWayMergePatch function hardcodes the `setElementOrder` diff option, which creates useless or undesired diffs

@Quentin-M I agree some people want to use SMP package and don't care about element order at all. So IMO we should give them option to opt out.

**Comment by mengqiy** [comment]
The code itself looks good. I will need to think more about backwards compatibility tomorrow.

File: staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go
Comment (mengqiy) [inline staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go]: This is not a nested map, please fix the description.
And we should have a few more test cases:
```
-   originalObj:
      mapKey:
           nonNilKey: {}
-   originalObj:
      mapKey:
           nonNilKey1: {}
           nonNilKey2: {}
-   originalObj:
      mapKey:
           nonNilKey: {}
           nilKey: null
```

File: staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go
Comment (mengqiy) [inline staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go]: Just found a bug in the test case.
We should use different keys:
```
         nonNilKey1: foo
         nonNilKey2: bar
```
There are other places with the same bug, please fix them.

**Review by mengqiy — COMMENTED** [review]
The code LGTM and it should not break backward compatibility.

File: staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go
Comment (davidhao3300) [reply staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go]: technically it's an empty map that's nested in another map, but I'll change to make it clearer

I'll add the suggested testcases

File: staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go
Comment (davidhao3300) [reply staging/src/k8s.io/apimachinery/pkg/util/jsonmergepatch/patch_test.go]: sounds good

**Comment by mengqiy** [comment]
/lgtm
Don't forget to squash the commit before merging.

/cc @liggitt

**Comment by mengqiy** [comment]
@lavalamp @liggitt Thoughts or concerns?

**Comment by davidhao3300** [comment]
Could I get an lgtm on this PR?

**Comment by lavalamp** [comment]
@davidhao3300 I agree with this code change and I'm willing to lgtm/approve. However! I read the issue that this is fixing and I do *not* want to give the impression that we are agreeing to keep that use case working in the future. Unfortunately, I intend to break that use case by the end of the year if at all possible. Apply is moving to the control plane; you'll need to come up with another permissions strategy. Please confirm that you've read this and I'll lgtm. Also, please edit & remove the "Fixes #58793" line from the PR description.

**Comment by davidhao3300** [comment]
As discussed in the issue linked above, I've agreed that this won't be necessarily supported in the long run

**Comment by lavalamp** [comment]
/approve
/lgtm

Thanks for the discussion @davidhao3300!

**Pull request merged**
